src/Compilers/Core/Portable/CodeAnalysisEventSource.cs

```diff
@@ -22,20 +22,35 @@ public static class Tasks
         {
             public const EventTask GeneratorDriverRunTime = (EventTask)1;
             public const EventTask SingleGeneratorRunTime = (EventTask)2;
+
+            public const EventTask PooledWhenSparse = (EventTask)3;
+            public const EventTask HalvedPooledCapacity = (EventTask)4;
         }
 
         private CodeAnalysisEventSource() { }
 
         [Event(1, Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Start, Task = Tasks.GeneratorDriverRunTime)]
-        internal void StartGeneratorDriverRunTime(string id) => WriteEvent(1, id);
+        internal void StartGeneratorDriverRunTime(string id)
+            => WriteEvent(1, id);
 
         [Event(2, Message = "Generators ran for {0} ticks", Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Stop, Task = Tasks.GeneratorDriverRunTime)]
-        internal void StopGeneratorDriverRunTime(long elapsedTicks, string id) => WriteEvent(2, elapsedTicks, id);
+        internal void StopGeneratorDriverRunTime(long elapsedTicks, string id)
+            => WriteEvent(2, elapsedTicks, id);
 
         [Event(3, Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Start, Task = Tasks.SingleGeneratorRunTime)]
-        internal void StartSingleGeneratorRunTime(string generatorName, string assemblyPath, string id) => WriteEvent(3, generatorName, assemblyPath, id);
+        internal void StartSingleGeneratorRunTime(string generatorName, string assemblyPath, string id)
+            => WriteEvent(3, generatorName, assemblyPath, id);
 
         [Event(4, Message = "Generator {0} ran for {2} ticks", Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Stop, Task = Tasks.SingleGeneratorRunTime)]
-        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id) => WriteEvent(4, generatorName, assemblyPath, elapsedTicks, id);
+        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id)
+            => WriteEvent(4, generatorName, assemblyPath, elapsedTicks, id);
+
+        [Event(5, Message = "Pooled when sparse: {0} {1}/{2}", Keywords = Keywords.Performance, Level = EventLevel.Informational, Task = Tasks.PooledWhenSparse)]
+        internal void PooledWhenSparseImpl(Type type, int count, int capacity, string id)
```

Comment:
The method names are used for the actual ETW events, so probably want to rename this to not contain `Impl`